### PR TITLE
Collect Spark apps

### PR DESCRIPTION
**Pull request by thorsteinnth**
Spark pod controller added
- Discovers Spark applications on the cluster and creates a Wave SparkApplication CR to represent them
- Includes information in CR about the driver pod and executor pods
- Calls out to Spark API to get more Spark specific data and adds it to CR
  - Calls Spark API on driver pod if the application is running, otherwise history server
- The CR is updated with new information troughout the lifetime of the Spark application

### CODE REVIEW

File: controllers/sparkpod_controller.go [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+)
+
+const (
+	SparkRoleLabel = "spark-role"
+	SparkAppLabel  = "spark-app-selector"
+	DriverRole     = "driver"
+	ExecutorRole   = "executor"
+
+	AppLabel                       = "app"
+	AppLabelValueEnterpriseGateway = "enterprise-gateway"
```
Comment (ntfrnzn) [inline controllers/sparkpod_controller.go:29]: semantics: would this be better as AppEnterpriseGatewayLabelValue?

File: controllers/sparkpod_controller.go [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+		}
+	} else {
+		err := r.createSparkApplicationCr(ctx, deepCopy)
+		if err != nil {
+			return false, fmt.Errorf("could not create spark application cr, %w", err)
+		}
+	}
+
+	// Set owner reference driver pod -> spark application CR if needed
+	shouldSetOwnerReference := shouldSetPodOwnerReference(pod, heritage)
```
Comment (ntfrnzn) [inline controllers/sparkpod_controller.go:230]: I think, fold the owner reference setting into createSparkApplicationCr() (add a pod argument to it) and requeue if changed.

**Review by ntfrnzn — COMMENTED** [review]
I left this till very late, and need to read further tomorrow.  Looks like good work.

File: controllers/sparkpod_controller.go [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+		Client:             client,
+		ClientSet:          clientSet,
+		getSparkApiManager: sparkApiManagerGetter,
+		Log:                log,
+		Scheme:             scheme,
+	}
+}
+
+// SparkApiManagerGetter is a factory function that returns an implementation of sparkapi.Manager
+type SparkApiManagerGetter func(clientSet kubernetes.Interface, driverPod *corev1.Pod, logger logr.Logger) (sparkapi.Manager, error)
```
[the remark this answers is not in the document]
Comment (thorsteinnth) [reply controllers/sparkpod_controller.go:66]: I like it. Good for testing and flexibility. In this case, the Spark API manager will talk either with the driver pod or the history server, depending on the current system status. This pattern deals with that nicely.

File: controllers/sparkpod_controller.go [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+			err := r.Client.Patch(ctx, deepCopy, client.MergeFrom(p))
+			if err != nil {
+				log.Error(err, "could not set finalizer")
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	shouldRequeue := false
+	switch sparkRole {
```
[the remark this answers is not in the document]
Comment (thorsteinnth) [reply controllers/sparkpod_controller.go:180]: I am not returning here so we can continue to the next block to remove the finalizer (after the final successful handleDriver/Executor)

File: controllers/sparkpod_controller.go (hunk 2 above)
Comment (thorsteinnth) [reply controllers/sparkpod_controller.go]: not sure about creating CR -> fetching CR (to get the UID) -> updating pod in the same method. Now we requeue after creating the CR, one level up.

**Review by ntfrnzn — APPROVED** [review]
👍

**Comment by ntfrnzn** [comment]
I still think the trident-style proxy-through-k8s requests are strange, but maybe that's just me ;)

**Pull request merged**